### DISCUSSION TITLE
How to propose jRuby to an Architect in 2022?

### DISCUSSION THREAD

**Discussion by spark08** [jruby/jruby#7155] [General]
Hi,

I am interested in a discussion on how to provide a case to start using a language within a Java based software development company. Proposal to an Architect in 2022 with the current state of popularity.

Primarily jRuby and for developing quick MVP and proof of concepts.

As a note, Oracle JDK and OpenJDK are being used, not GraalVM

Thanks

**Comment by mohits** [comment]
Hi @spark08 - what kind of information or discussion are you looking for? I had been meaning to write a post about why we use JRuby:
- Runs where we have Java but not Ruby
- Faster for multiple use cases
- Packages & distributes easily as a JAR
- Connects to Java libraries
- Fantastic people

This is now at: https://notepad.onghu.com/2022/why-we-use-jruby/

What can we help put together?

**Reply by spark08** [reply]
Hi @mohits 

Very nice writeup, enjoyed reading alot. 

My question is more about the business values of jRuby to a Java shop. Why would a company start implementing and supporting this new language. Some first interactions with an Architect showed that his impression is that jRuby/Ruby is very unpopular.

Thoughts?

**Reply by mohits** [reply]
Hi @spark08 - thanks for the kind words on the post.

As I noted in the post, we are not a Java shop - but from the outside, I would imagine the main advantages being:
1. Speed of development, especially for POCs when using things like Rails, ActionMailer, ActiveRecord, Sequel, Roda, etc. Also, these are production ready and can be deployed to a customer even if there is a plan to rewrite these into Java eventually.
2. Possibly rapid iteration on ideas using Java libraries and exploring/ trying things out with them
3. Using something like RSpec for running tests on Java libraries or systems (we recently used RSpec for doing a whole battery of tests on a C++ executable). We found this to be much more productive than doing tests using a C++ based framework.

Of course, all of these are predicated on being familiar with Ruby and wanting to use Ruby stuff as a means to accelerate development. Also, some of these (esp. Items 3 and 1) could be done in Ruby directly without needing JRuby.

Hope this helps. I'm sure @headius  has more ideas on what else could be used.

**Reply by mohits** [reply]
Also on this:

> jRuby/Ruby is very unpopular

I don't know what to say - generally speaking, Ruby is regularly in the top 20 languages and for web backend development, probably in the Top 10 (or even Top 5). Also, Ruby is used for GitHub and Shopify, some things from Apple, and many other large web properties.

At the end, it would go down to what they want to achieve, and whether Ruby is a sensible tool in the stack (e.g. familiarity among developers, etc.)

**Reply by spark08** [reply]
Hi @mohits 

Thank you. Point 3 is very interesting, especially since this shop uses also C++ applications. I did not think about it, I'll use it.

Points 1 & 2 go hand in hand and it's very true. Comparing estimates for delivering functionality in Java vs Ruby, Ruby clearly has advantage here.

I agree with you on popularity statement. Unfortunately, only around 3% of the developers in this shop heard or used ruby, based on their CV. Regardless, I still want to build a solid case to start using Ruby. Due to points 2&3 you mentioned above.

I thank you once more for your insights, they were helpful

**Reply by mattwelke** [reply]
Tech recommendations are hard as hell. The way I've noticed to make it well is to wait til you've got reputation at your org to start making the recommendations, and come to them with a clear explanation of how it's going to benefit. Examples don't hurt either. For JSON Schema (instead of JSON going over APIs, without a schema), I showed off code generation tools, and the ability to document inline. For merging microservices into monoliths, I explained how the workloads were similar and that one wouldn't impact the other. Side benefit is once you've got an example, you can put it in a public GitHub repo and build up a portfolio that way. Two birds with one stone. :P
